design/open/design-datasets.md

```diff
@@ -0,0 +1,221 @@
+  support (HMS is the catalog).  Our Iceberg support _might_ become partly
+  external, if we added support to defer table definitions to some other
+  catalog.
+* Internally, by effectively serving the catalog.  This is our lakeFS Delta
+  Sharing (the Delta Sharing protocol does not call this a "catalog").  Our
+  Iceberg support is also currently internal.
+
+### Implicit definitions
+
+Implicit definitions give users the _easiest_ path.  However they have
+multiple limitations:
+
+* Require discoverability of the table by examining only object metadata.
+  So supported only by "modern" table formats: Iceberg and Delta.  Formats
+  such as prefixes containing Parquet files _cannot_ support all operations:
+  - It is not possible to report table metadata efficiently.
+  - It may not be possible to list all tables.
+  - It will _not_ be possible to support named tables, all tables must be
+    defined by paths.
+
+Using implicit definitions necessarily leaks through the table abstraction.
+There is no useful general case here.  Users will find it easy to request
+functionality that uses implicit definitions.
+
+#### Possible future features
+
+Define an implicit "Parquet folder" as a folder that holds many `.parquet`
+files, or a folder that holds many `<key>=<value>` subfolders with some
+`.parquet` files under them.  Then we could add to the GUI a display of a
+sample DuckDB query when looking at a Parquet folder:
+
+### External definitions
+
+External definitions in catalogs may be the easiest way to give _some_ users
+a fully-functional experience, for _some_ catalogs.  Their principal
+limitations:
+
+* Risk of fragmentation: If there are many popular external catalogs we will
+  need to support many of them.  Different catalogs support different
+  use-cases, so we should expect to have to support multiple catalogs.
+* Limitations due to the external definition: Because existing catalog
+  implementations were designed without lakeFS in mind, we will not be able
+  to add lakeFS features to all of them.  For instance the HMS protocol is
+  Thrift-based and does not support refs-as-schemas.
+
+Iceberg support might become an external definition once it supports
+deferring to an external catalog.  It does not yet support wrapping an
+external catalog because of these risks.
+
+#### Possible future features
+
+Support more catalog types.  Each catalog adds implementation effort.
+
+Support catalog wrappers that work with multiple external catalogs.  For
+instance, wrapping additional catalogs under the Iceberg catalog, making it
+an external definition.
+
+### Internal definition
+
+An internal definition should allow us to support _all_ lakeFS features that
+systems can handle.  Because it reduces external dependencies, it is
+probably the _fastest_ way for us to implement "table" features on top of
+lakeFS -- we have seen this in the lakeFS Delta Sharing server.  The Iceberg
```

Comment:
I'm still not sure that the Iceberg catalog is considered internal. I agree that the catalog's files are stored using lakeFS, yet the implementation is completely external to lakeFS...

Reply:
As a catalog, we are internal: we "serve" this catalog.  The implementation leverages another implementation.  But mapping of names to locations goes through our catalog.  (It doesn't do much with it, but it could.)